### PR TITLE
UCP/WORKER: Fix worker cleanup order

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by brminich** [comment]
error is relevant:
```
#                  put[1]: 0..<short>..4294967296..<bcopy>..(inf)
#                  get[1]: 0..<bcopy>..(inf)
#                  put[2]: 0..<short>..4294967296..<bcopy>..(inf)
#                  get[2]: 0..<bcopy>..(inf)
#
#                  rma_bw: mds [8] [12] rndv_rkey_size 35
#
[swx-rdmz-ucx-gpu-01:30601:0:30601] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xc8)

/scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucs/type/spinlock.h: [ ucs_recursive_spin_lock() ]
      ...
      107 static inline void ucs_recursive_spin_lock(ucs_recursive_spinlock_t *lock)
      108 {
      109     pthread_t self = pthread_self();
==>   110 
      111     if (ucs_recursive_spin_is_owner(lock, self)) {
      112         ++lock->count;
      113         return;

==== backtrace (tid:  30601) ====
 0 0x0000000000053325 ucs_debug_print_backtrace()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucs/debug/debug.c:656
 1 0x000000000004c466 ucs_recursive_spin_lock()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucs/type/spinlock.h:110
 2 0x000000000004c466 ucs_callbackq_enter()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucs/datastruct/callbackq.c:62
 3 0x000000000004c466 ucs_callbackq_remove_if()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucs/datastruct/callbackq.c:555
 4 0x00000000000189ea ucp_ep_delete()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_ep.c:168
 5 0x00000000000189ea ucp_ep_delete()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_ep.c:171
 6 0x0000000000018fbd ucp_worker_destroy_mem_type_endpoints()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_ep.c:309
 7 0x00000000000258d2 ucp_worker_destroy()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_worker.c:2062
 8 0x00000000000258d2 ucp_worker_close_cms()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_worker.c:1349
 9 0x00000000000258d2 ucp_worker_destroy()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_worker.c:2063
10 0x000000000040267f print_ucp_info()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/tools/info/proto_info.c:206
11 0x00000000004021ed main()  /scrap/azure/agent-04/AZP_WORKSPACE/2/s/contrib/../src/tools/info/ucx_info.c:215
12 0x00000000000223d5 __libc_start_main()  ???:0
13 0x00000000004022a3 _start()  ???:0
```

**Comment by brminich** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@evgeny-leksikov can you pls take a look?

File: src/ucp/core/ucp_worker.c [hunk 1]
```diff
@@ -2021,27 +2064,24 @@ void ucp_worker_destroy(ucp_worker_h worker)
...
-    kh_destroy_inplace(ucp_worker_rkey_config, &worker->rkey_config_hash);
-    ucs_ptr_map_destroy(&worker->ptr_map);
-    ucs_strided_alloc_cleanup(&worker->ep_alloc);
     UCS_STATS_NODE_FREE(worker->tm_offload_stats);
     UCS_STATS_NODE_FREE(worker->stats);
+    ucs_ptr_map_destroy(&worker->ptr_map);
+    ucs_strided_alloc_cleanup(&worker->ep_alloc);
+    kh_destroy_inplace(ucp_worker_rkey_config, &worker->rkey_config_hash);
+    ucp_worker_destroy_ep_configs(worker);
+    ucs_conn_match_cleanup(&worker->conn_match_ctx);
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c:2084]: logically `ucs_conn_match_cleanup` should be done before EPs and ptr_map clean up

File: src/ucp/core/ucp_worker.c [hunk 2]
```diff
@@ -1918,40 +1957,44 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
...
 
+    /* Initialize connection matching structure */
+    ucs_conn_match_init(&worker->conn_match_ctx, sizeof(uint64_t),
+                        &ucp_ep_match_ops);
+
     /* Open all resources as connection managers on this worker */
     status = ucp_worker_add_resource_cms(worker);
     if (status != UCS_OK) {
-        goto err_close_cms;
+        goto err_close_ifaces;
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c:1973]: `goto err_conn_match_cleanup;`

**Review by evgeny-leksikov — APPROVED** [review]

**Comment by yosefe** [comment]
@dmitrygx @evgeny-leksikov can you pls look again (fixed conflicts with discard ep + small fix for discard hash cleanup)

**Review by evgeny-leksikov — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

**Pull request merged**